File: app/src/main/java/org/tasks/notifications/Throttle.java
package org.tasks.notifications;

import static com.todoroo.andlib.utility.AndroidUtilities.assertNotMainThread;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;

class Throttle {
  private final long[] throttle;
  private final Sleeper sleeper;
  private int oldest = 0;

  Throttle(int ratePerSecond) {
    this(ratePerSecond, Throttle::sleep);
  }

  Throttle(int ratePerSecond, Sleeper sleeper) {
    this.sleeper = sleeper;
    throttle = new long[ratePerSecond];
  }

  private static void sleep(long millis) {
    try {
      Thread.sleep(millis);
    } catch (InterruptedException ignored) {
    }
  }

  synchronized void run(Runnable runnable) {
    assertNotMainThread();
    long sleep = throttle[oldest] - (currentTimeMillis() - 1000);
    if (sleep > 0) {
      sleeper.sleep(sleep);
    }
    runnable.run();
    throttle[oldest] = currentTimeMillis();
    oldest = (oldest + 1) % throttle.length;
  }

  public interface Sleeper {
    void sleep(long millis);
  }
}

File: app/src/main/java/org/tasks/notifications/Throttle.kt
package org.tasks.notifications

import kotlinx.coroutines.runBlocking
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import timber.log.Timber
import java.util.concurrent.Executor
import java.util.concurrent.Executors.newSingleThreadExecutor

internal class Throttle(
        ratePerPeriod: Int,
        private val periodMillis: Long = 1000,
        private val tag: String = "",
        private val executor: Executor = newSingleThreadExecutor(),
        private val sleeper: (Long) -> Unit = { Thread.sleep(it) }) {
    private val throttle: LongArray = LongArray(ratePerPeriod)
    private var oldest = 0

    @Synchronized
    fun run(runnable: suspend () -> Unit) {
        executor.execute {
            val sleep = throttle[oldest] - (currentTimeMillis() - periodMillis)
            if (sleep > 0) {
                Timber.v("$tag: Throttled for ${sleep}ms")
                sleeper(sleep)
            }
            try {
                runBlocking {
                    runnable()
                }
            } catch (e: Exception) {
                Timber.e(e)
            }
            throttle[oldest] = currentTimeMillis()
            oldest = (oldest + 1) % throttle.size
        }
    }

    fun pause(millis: Long) = executor.execute {
        Thread.sleep(millis)
    }
}
